cli/snapshot.go

```diff
@@ -0,0 +1,169 @@
+package cli
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// ioctl magics from <linux/fs.h>
+	IOCTL_FIFREEZE_MAGIC = 0xC0045877 // _IOWR('X', 119, int)
+	IOCTL_FITHAW_MAGIC   = 0xC0045878 // _IOWR('X', 120, int)
+)
+
+type SnapshotOptions struct {
+	ArtifactName      string
+	ArtifactGroup     string
+	CompatibleDevices []string
+	Depends           []string
+	Provides          []string
+}
+
+type mountInfo struct {
+	mountID    int
+	mountPoint string
+	mountSrc   string
+	fsType     string
+}
+
+func (runOpts *runOptionsType) DumpSnapshot(ctx *cli.Context) error {
+	rootDev, err := getFsDevFile("/")
+	if err != nil {
+		return err
+	}
+	dataDev, err := getFsDevFile(runOpts.dataStore)
+	if err != nil {
+		return err
+	}
+	if rootDev == dataDev {
+		return errors.Errorf(
+			"State data store (%s) is located on rootfs partition",
+			runOpts.dataStore)
+	}
+
+	errChan := make(chan error)
+	go func() {
+		thawChan := make(chan int)
+		sigChan := make(chan os.Signal)
+		signal.Notify(sigChan, unix.SIGINT, unix.SIGPIPE)
+		go func() {
+			select {
+			case <-sigChan:
+			case <-thawChan:
+			}
+			// FIXME: handle error
+			thawFs("/")
+			// Invoke os signal handler
+			signal.Stop(sigChan)
+			unix.Kill(os.Getpid(), unix.SIGINT)
+		}()
+		freezeFs("/")
+		unix.Sync()
```

Comment:
Will this work? I would expect this to potentially block indefinitely after `freezeFs` has been called. This sequence of events is fine:

```
freezeFs("/")
unix.Sync()
--CONTEXT SWITCH--
<Some other process calls write()
```

However, if you're unlucky, I think this next one will block forever:

```
freezeFs("/")
--CONTEXT SWITCH--
<Some other process calls write()
--CONTEXT SWITCH--
unix.Sync()
```

I think `Sync()` should not be necessary though. Digging into the kernel source, it appears that [`Sync()` is called for you, as part of the freeze](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/super.c#n1729), and I assume this is atomic, which is better than you can hope for by calling it manually.

Reply:
Ohh, I was confused after I read somewhere that all writes *after* calling `FIFREEZE` would be blocked, so I added a Sync to make sure that all cached writes would be flushed (for good measure). But I see I should probably consolidated with the sources first :)